---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# DEGSimilar

<!-- badges: start -->
<!-- badges: end -->

The goal of DEGSimilar is to provide an easy pipeline in order to analyze the results of Differential Gene Expression (DEG).
This is done by providing an easy method to compare DEGs of choice to other DEGs based on BLAST similiarity within a user-specified
database.

## Installation

Yout can install the package from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("GeorgeLy7/DEGSimilar")
```

In order to use this package the BLAST+ executable must be installed. It may be installed from here:
https://blast.ncbi.nlm.nih.gov/Blast.cgi?CMD=Web&PAGE_TYPE=BlastDocs&DOC_TYPE=Download

## Example

This is a basic example which shows you how to solve a common problem:

```{r example}
library(DEGSimilar)
blastResults <- getBlastResults("./data-raw/newTest2.fasta","./data-raw/StHeBC3.fasta")
blastResults
```

You'll still need to render `README.Rmd` regularly, to keep `README.md` up-to-date.

You can also embed plots, for example:

```{r pressure, echo = FALSE}
plot(pressure)
```

In that case, don't forget to commit and push the resulting figure files, so they display on GitHub!
